### PR TITLE
Adds a code of conduct adapted from the Rust CoC

### PR DESCRIPTION
**Pull request by fhaynes**
We've needed one of these for awhile. I modified it to fit our project. For now, I put myself and torkleyy down as moderators, as I'm not sure who else is interested.

### CODE REVIEW

**Comment by distransient** [comment]
I really like this! As long as it's something the core team is happy with, especially in terms of moderators I'd really like it to be merged.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+1. Remarks that violate the Amethyst standards of conduct, including hateful, hurtful, oppressive, or exclusionary remarks, are not allowed. (Cursing is allowed, but never targeting another user, and never in a hateful manner.)
+2. Remarks that moderators find inappropriate, whether listed in the code of conduct or not, are also not allowed.
+3. Moderators will first respond to such remarks with a warning.
+4. If the warning is unheeded, the user will be "kicked," i.e., kicked out of the communication channel to cool off.
+5. If the user comes back and continues to make trouble, they will be banned, i.e., indefinitely excluded.
+6. Moderators may choose at their discretion to un-ban the user if it was a first offense and they offer the offended party a genuine apology.
+7. If a moderator bans someone and you think it was unjustified, please take it up with that moderator, or with a different moderator, **in private**. Complaints about bans in-channel are not allowed.
+8. Moderators are held to a higher standard than other community members. If a moderator creates an inappropriate situation, they should expect less leeway than others.
```
Comment (torkleyy) [inline CODE_OF_CONDUCT.md]: Let's add organization members to this.

File: CODE_OF_CONDUCT.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+* On Discord, the forums, IRC, GitHub, or any other community area, please avoid using overtly sexual nicknames or other nicknames that might detract from a friendly, safe and welcoming environment for all.
+* Please be kind and courteous. There's no need to be mean or rude.
+* Respect that people have differences of opinion and that every design or implementation choice carries a trade-off and numerous costs. There is seldom a right answer.
+* Please keep unstructured critique to a minimum. If you have solid ideas you want to experiment with, make a fork and see how it works.
+* We will exclude you from interaction if you insult, demean or harass anyone. That is not welcome behavior. We interpret the term "harassment" as including the definition in the <a href="http://citizencodeofconduct.org/">Citizen Code of Conduct</a>; if you have any lack of clarity about what might be included in that concept, please read their definition. In particular, we don't tolerate behavior that excludes people in socially marginalized groups.
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: This link seems to be down for me.

Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: Also it should probably be markdown, right?

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+* Private harassment is also unacceptable. No matter who you are, if you feel you have been or are being harassed or made uncomfortable by a community member, please contact one of the channel ops or any of the [Amethyst moderation team][mod_team] immediately. Whether you're a regular contributor or a newcomer, we care about making this community a safe place for you and we've got your back.
+* Likewise any spamming, trolling, flaming, baiting or other attention-stealing behavior is not welcome.
+
+## Moderation
+
+
+These are the policies for upholding our community's standards of conduct. If you feel that a thread needs moderation, please contact the [Amethyst moderation team][mod_team].
+
+1. Remarks that violate the Amethyst standards of conduct, including hateful, hurtful, oppressive, or exclusionary remarks, are not allowed. (Cursing is allowed, but never targeting another user, and never in a hateful manner.)
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: This makes it seem almost like we encourage cursing to some degree. If possible, I would like to add something saying that it's generally discouraged.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: I would say that we should probably just say "anyone in close relation to the Amethyst Foundation" which would cover all possible scenarios.

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+8. Moderators are held to a higher standard than other community members. If a moderator creates an inappropriate situation, they should expect less leeway than others.
+
+In the Amethyst community we strive to go the extra step to look out for each other. Don't just aim to be technically unimpeachable, try to be your best self. In particular, avoid flirting with offensive or sensitive issues, particularly if they're off-topic; this all too often leads to unnecessary fights, hurt feelings, and damaged trust; worse, it can drive people away from the community entirely.
+
+And if someone takes issue with something you said or did, resist the urge to be defensive. Just stop doing what it was they complained about and apologize. Even if you feel you were misinterpreted or unfairly accused, chances are good there was something you could've communicated better — remember that it's your responsibility to make your fellow Rustaceans comfortable. Everyone wants to get along and we are all here first and foremost because we want to talk about cool technology. You will find that people will be eager to assume good intent and forgive as long as you earn their trust.
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: This paragraph needs a bit of rewording to adapt to Amethyst.

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+The enforcement policies listed above apply to all official Amethyst venues; including all Amethyst Discord channels; GitHub repositories under amethyst; and the forums at community.amethyst-engine.org.
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: I think it would be nice to say something like "feel free to link to this CoC if you want your own project to follow these rules".

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+*Adapted from the [Node.js Policy on Trolling](http://blog.izs.me/post/30036893703/policy-on-trolling) as well as the [Contributor Covenant v1.3.0](https://www.contributor-covenant.org/version/1/3/0/).*
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: I don't like that we have some inline links and some are aliases. Could we keep it consistent maybe?

File: CODE_OF_CONDUCT.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# The Amethyst Code of Conduct
+
+## Conduct
+
+**Contact**: [amethyst-mods@amethyst-engine.org](mailto:amethyst-mods@amethyst-engine.org)
+
+* We are committed to providing a friendly, safe and welcoming environment for all, regardless of level of experience, gender identity and expression, sexual orientation, disability, personal appearance, body size, race, ethnicity, age, religion, nationality, or other similar characteristic.
+* On Discord, the forums, IRC, GitHub, or any other community area, please avoid using overtly sexual nicknames or other nicknames that might detract from a friendly, safe and welcoming environment for all.
+* Please be kind and courteous. There's no need to be mean or rude.
+* Respect that people have differences of opinion and that every design or implementation choice carries a trade-off and numerous costs. There is seldom a right answer.
```
Comment (jojolepro) [inline CODE_OF_CONDUCT.md]: >seldom 

I didn't know that word until today :P

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (jojolepro) [inline CODE_OF_CONDUCT.md]: > In particular, we don't tolerate behavior that excludes people in socially marginalized groups.

Should be:
"In particular, we don't tolerate behavior that excludes people."

Its pretty easy to think about the big group hitting the smaller group, but the opposite is also true.

File: CODE_OF_CONDUCT.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+* Private harassment is also unacceptable. No matter who you are, if you feel you have been or are being harassed or made uncomfortable by a community member, please contact one of the channel ops or any of the [Amethyst moderation team][mod_team] immediately. Whether you're a regular contributor or a newcomer, we care about making this community a safe place for you and we've got your back.
+* Likewise any spamming, trolling, flaming, baiting or other attention-stealing behavior is not welcome.
+
+## Moderation
+
+
+These are the policies for upholding our community's standards of conduct. If you feel that a thread needs moderation, please contact the [Amethyst moderation team][mod_team].
+
+1. Remarks that violate the Amethyst standards of conduct, including hateful, hurtful, oppressive, or exclusionary remarks, are not allowed. (Cursing is allowed, but never targeting another user, and never in a hateful manner.)
+2. Remarks that moderators find inappropriate, whether listed in the code of conduct or not, are also not allowed.
```
Comment (jojolepro) [inline CODE_OF_CONDUCT.md]: That is really vague and easily abusable imo

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (jojolepro) [inline CODE_OF_CONDUCT.md]: If a moderator **voluntarily** creates an inappropriate situation

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (jojolepro) [reply CODE_OF_CONDUCT.md]: I think it made a lot of sense.

File: CODE_OF_CONDUCT.md (hunk 8 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: To some extent, it's supposed to be that way. Being somewhat unspecific about it means that people will be a bit more careful when they interact with others. Of course, that falls apart quickly if the moderation team isn't any good but I don't see that becoming a problem in Amethyst anytime soon.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: No. If they created one without having any clue what they're doing, they shouldn't be a moderator. A moderator **definitely** shouldn't be able to escape with "oh sorry I didn't mean it"

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: It's not about it not making sense, it's about it feeling like it was ripped straight off of Rust's CoC 😛

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (jojolepro) [reply CODE_OF_CONDUCT.md]: If I say something in chat which is perfectly accepted in canada, but is deemed racist in another country and I didn't know. Someone calls me out and I apologize. In this situation, what happens?

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: It was ripped straight off of Rust's CoC. =P

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: Rust's original CoC has that, which I removed, since this is a modified version. I can add something noting that and linking to the original rules.

File: CODE_OF_CONDUCT.md (hunk 6 above)
Comment (Moxinilian) [inline CODE_OF_CONDUCT.md]: We should also state that it is adapted from Rust's CoC.

File: CODE_OF_CONDUCT.md (hunk 7 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: It is seldom used

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: I am not comfortable changing this. We wouldn't tolerate hate groups, for example. One of the core parts of Rust's community is that it, and the community of many Rust projects, pro-actively take steps to protect socially marginalized groups that are historically demonized or threatened online or in real-life.

Comment (Moxinilian) [reply CODE_OF_CONDUCT.md]: But we wouldn't tolerate hate group because as stated elsewhere we don't tolerate hatefulness. I think we already cover elsewhere the exceptions to that principle.

File: CODE_OF_CONDUCT.md (hunk 8 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: Having moderated large communities before, creating dozens of micro-rules just gives rules lawyers things to argue about.

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: Site is down it seems. Or at least their DB server. I removed that text for now.

File: CODE_OF_CONDUCT.md (hunk 3 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: How's this:

> (Cursing in general is discouraged, and never tolerated if targeting another user, and never in a hateful manner.)

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: That's what I mean! Adapt it more! :smile: We're supposed to include everyone, right? What if a Lua developer starts engaging with the Amethyst community and never learns Rust, they wouldn't be a rustacean, would they? Does that mean they're excluded from the rules specified here? I mean, obviously not, but I think you get my point :stuck_out_tongue:

Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: I'm unsure how to make it more applicable to Amethyst. Ideas?

File: CODE_OF_CONDUCT.md (hunk 6 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: mod_team is reference linked because it is used multiple places in the document. The others are more footnotes. Do you mean the 

> the forums at community.amethyst-engine.org

part?

File: CODE_OF_CONDUCT.md (hunk 3 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: Sounds good to me :smile:

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]:
```suggestion
And if someone takes issue with something you said or did, resist the urge to be defensive. Just stop doing what it was they complained about and apologize. Even if you feel you were misinterpreted or unfairly accused, chances are good there was something you could've communicated better — remember that it's your responsibility to make your fellow Amethyst community members comfortable. Everyone wants to get along and we are all here first and foremost because we want to talk about making games. You will find that people will be eager to assume good intent and forgive as long as you earn their trust.
```

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: Ah okay. I'm thinking that this one may be slightly more appropriate for game development projects but your call I suppose :wink:

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: I am, however, ok adding something like "We strive to be an inclusive community"

Comment (Moxinilian) [reply CODE_OF_CONDUCT.md]: I don't think it is necessary to be so specific. I think it is pretty clear what we are talking about, and so can be as general as it gets.

File: CODE_OF_CONDUCT.md (hunk 6 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: Done

**Comment by fhaynes** [comment]
@magnonellie @jojolepro @LucioFranco (and anyone else) I made changes, please review them.

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (Moxinilian) [reply CODE_OF_CONDUCT.md]: What do you mean?

Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: @Moxinilian I'm not sure what you're asking. Can you rephrase?

File: CODE_OF_CONDUCT.md (hunk 8 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: A good team of moderators will not abuse this, and if one does, the rest of the moderation team will deal with it immediately. What I've seen in the past is people using "But it isn't in the code of conduct!" as an excuse, and then every possible inappropriate thing gets added to a huge list.

I'm open to re-phrasing that part, but will need some suggestions.

File: CODE_OF_CONDUCT.md (hunk 6 above)
Comment (magnonellie) [reply CODE_OF_CONDUCT.md]: I mean just in general I would like to have everything defined in the bottom like we do in most other markdown documents.

File: CODE_OF_CONDUCT.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+And if someone takes issue with something you said or did, resist the urge to be defensive. Just stop doing what it was they complained about and apologize. Even if you feel you were misinterpreted or unfairly accused, chances are good there was something you could've communicated better — remember that it's your responsibility to make your fellow Amethyst community members comfortable. Everyone wants to get along and we are all here first and foremost because we want to talk about making games. You will find that people will be eager to assume good intent and forgive as long as you earn their trust.
+
+The enforcement policies listed above apply to all official Amethyst venues; including all Amethyst Discord channels; GitHub repositories under amethyst; and the community forums at.
+
+This Code of Conduct was based on the Rust Code of Conduct, available here: [https://www.rust-lang.org/conduct.html]
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: This sounds somewhat clunky having another "based on" section below. Maybe we can just remove the mentions below as they're pretty controversial anyway?

File: CODE_OF_CONDUCT.md (hunk 2 above)
Comment (Moxinilian) [reply CODE_OF_CONDUCT.md]: I think jojo's suggestion is legitimate for the reasons I stated above.

File: CODE_OF_CONDUCT.md (hunk 8 above)
Comment (Xaeroxe) [reply CODE_OF_CONDUCT.md]: I'd rephrase this to "Remarks that moderators find inappropriate, whether listed in the code of conduct or not, will be subject to the following:"

I think the phrase "not allowed." implies harsher punishment than intended.  We detail what exactly the consequences are pretty much right away, but we kind of get off on the wrong foot here.

File: CODE_OF_CONDUCT.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+In the Amethyst community we strive to go the extra step to look out for each other. Don't just aim to be technically unimpeachable, try to be your best self. In particular, avoid flirting with offensive or sensitive issues, particularly if they're off-topic; this all too often leads to unnecessary fights, hurt feelings, and damaged trust; worse, it can drive people away from the community entirely.
+
+And if someone takes issue with something you said or did, resist the urge to be defensive. Just stop doing what it was they complained about and apologize. Even if you feel you were misinterpreted or unfairly accused, chances are good there was something you could've communicated better — remember that it's your responsibility to make your fellow Amethyst community members comfortable. Everyone wants to get along and we are all here first and foremost because we want to talk about making games. You will find that people will be eager to assume good intent and forgive as long as you earn their trust.
```
Comment (Xaeroxe) [inline CODE_OF_CONDUCT.md]: While I generally feel a lot of what's given here is good advice, it's possibly written a tad too heavily in favor of the offended.  Communication is a two way street, and while it's important that we stress that people should do their part to make that work it's also valuable to allow the possibility that the offended is either acting in bad faith or just being completely unreasonable.  We don't need to say that explicitly here, but perhaps we should make this paragraph a bit less absolutist.  I'm open to additional discussion on this, but that's my first thoughts upon reading this.

Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: Can you give an example of how you would re-word it?

File: CODE_OF_CONDUCT.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+The enforcement policies listed above apply to all official Amethyst venues; including all Amethyst Discord channels; GitHub repositories under amethyst; and the community forums at [https://community.amethyst-engine.org]
+
+This Code of Conduct is based on the Rust Code of Conduct, available here: [https://www.rust-lang.org/conduct.html]
```
Comment (magnonellie) [inline CODE_OF_CONDUCT.md]: I don't think this link actually parses. Also I do still think italics could be appropriate here :wink:

File: CODE_OF_CONDUCT.md (hunk 10 above)
Comment (fhaynes) [reply CODE_OF_CONDUCT.md]: I spoke with @Xaeroxe and we're going to revisit this when he has more time to collaborate on a some different wording.
